onadata/apps/api/viewsets/user_profile_viewset.py

```diff
@@ -200,3 +213,84 @@ def monthly_submissions(self, request, *args, **kwargs):
 
         serializer = MonthlySubmissionsSerializer(instance_count, many=True)
         return Response(serializer.data[0])
+
+    @action(detail=False)
+    def verify_email(self, request, *args, **kwargs):
+        verified_key_text = getattr(settings, "VERIFIED_KEY_TEXT", None)
+
+        if not verified_key_text:
+            return Response(status=status.HTTP_204_NO_CONTENT)
+
+        redirect_url = request.query_params.get('redirect_url')
+        verification_key = request.query_params.get('verification_key')
+        response_message = _("Missing or invalid verification key")
+        if verification_key:
+            try:
+                rp = RegistrationProfile.objects.get(
+                    activation_key=verification_key
+                )
+            except RegistrationProfile.DoesNotExist:
+                pass
+            else:
+                rp.activation_key = verified_key_text
+                rp.save()
+
+                set_is_email_verified(rp.user.profile, True)
+
+                response_data = {
+                    'ona_verified_username': rp.user.username,
+                    'ona_verified_email_status': True
+                }
+
+                if redirect_url:
+                    query_params_string = urlencode(response_data)
+                    redirect_url = '{}?{}'.format(
+                        redirect_url, query_params_string
+                    )
+
+                    return HttpResponseRedirect(redirect_url)
+
+                return Response(response_data)
+
+        return HttpResponseBadRequest(response_message)
+
+    @action(methods=['POST'], detail=False)
+    def send_verification_email(self, request, *args, **kwargs):
+        verified_key_text = getattr(settings, "VERIFIED_KEY_TEXT", None)
+        if not verified_key_text:
+            return Response(status=status.HTTP_204_NO_CONTENT)
+
+        username = request.data.get('username')
+        redirect_url = request.data.get('redirect_url')
```

Comment:
The rest framework way I believe is we put this functionality in a serializer, not a big deal. Plenty of examples in this codebase.